bundles/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/discovery/NetworkDiscoveryService.java

```diff
@@ -120,52 +121,93 @@ public void partialDetectionResult(PresenceDetectionValue value) {
+                    t.setDaemon(true);
+                    return t;
+                }, new ThreadPoolExecutor.CallerRunsPolicy() // backpressure when saturated
+        );
+    }
+
     /**
      * Starts the DiscoveryThread for each IP on each interface on the network
      */
     @Override
     protected void startScan() {
         if (executorService == null) {
-            executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2);
+            executorService = createDiscoveryExecutor();
         }
         final ExecutorService service = executorService;
         if (service == null) {
             return;
         }
+
         removeOlderResults(getTimestampOfLastScan(), null);
-        logger.trace("Starting Network Device Discovery");
-
-        final Set<String> networkIPs = networkUtils.getNetworkIPs(MAXIMUM_IPS_PER_INTERFACE);
-        scannedIPcount.set(0);
-
-        for (String ip : networkIPs) {
-            final PresenceDetection pd = new PresenceDetection(this, scheduler, Duration.ofSeconds(2));
-            pd.setHostname(ip);
-            pd.setIOSDevice(true);
-            pd.setUseDhcpSniffing(false);
-            pd.setTimeout(PING_TIMEOUT);
-            // Ping devices
-            pd.setUseIcmpPing(true);
-            pd.setUseArpPing(true, configuration.arpPingToolPath, configuration.arpPingUtilMethod);
-            // TCP devices
-            pd.setServicePorts(tcpServicePorts);
-
-            service.execute(() -> {
-                Thread.currentThread().setName("Discovery thread " + ip);
-                try {
-                    pd.getValue();
-                } catch (ExecutionException | InterruptedException e) {
-                    stopScan();
-                }
-                int count = scannedIPcount.incrementAndGet();
-                if (count == networkIPs.size()) {
-                    logger.trace("Scan of {} IPs successful", scannedIPcount);
-                    stopScan();
-                }
-            });
+        logger.debug("Starting Network Device Discovery");
+
+        Map<String, Set<CidrAddress>> discoveryList = networkUtils.getNetworkIPsPerInterface();
+
+        // Track completion for all interfaces
+        final int totalInterfaces = discoveryList.size();
+        final AtomicInteger completedInterfaces = new AtomicInteger(0);
+
+        for (String networkInterface : discoveryList.keySet()) {
+            final Set<String> networkIPs = networkUtils.getNetworkIPs(
+                    Objects.requireNonNull(discoveryList.get(networkInterface)), MAXIMUM_IPS_PER_INTERFACE);
+            logger.debug("Scanning {} IPs on interface {} ", networkIPs.size(), networkInterface);
+            final AtomicInteger scannedIPcount = new AtomicInteger(0);
+
+            for (String ip : networkIPs) {
+                service.execute(() -> {
+                    PresenceDetection pd = presenceDetectorThreadLocal.get();
```

Comment:
I'm wondering what the purpose of using `ThreadLocal` is. These threads are executor threads, and they get a new `Runnable` object each time they execute. So, the `ThreaLocal` is GC'ed with the `Runnable` when the task is done, the next task will create a new one.

I don't think there's anytime to "save" by using `ThreadLocal` here - on the contrary, I think it just complicates everything.

Reply:
It was introduced as part of attempts to improve the performance. Before every scan created a new PresenceDetection instance. Each instance also initialized its internal state (TCP port list, ping settings, etc.), which adds cost beyond the raw allocation.
With the threadlocal one instance per worker thread is created and reused across all IP's. Less allocations, Less GC pressure. Less cpu cycles.

The PR changed over time and now i look at it again, and with your insights, there is not much to gain and it looks a micro-optimization. But yeah, you try something when chasing performance problems. :-/

Reply:
Reverted the usage of theadlocal